src/data/polynomial/hasse_deriv.lean

```diff
@@ -0,0 +1,220 @@
+    all_goals { simp only [forall_const, monomial_zero_right, mul_zero], },
+  end,
+  map_smul' := λ c f,
+  begin
+    rw [sum_eq_of_subset _ _ _ _ (support_smul c f), sum_def, finset.smul_sum,
+      finset.sum_congr rfl],
+    { intros i hi,
+      have := nat.cast_commute (i.choose k) c,
+      simp only [coeff_smul, smul_monomial, smul_eq_mul, ← mul_assoc, this.eq], },
+    { simp only [forall_const, monomial_zero_right, mul_zero], }
+  end }
+.
+
+lemma hasse_deriv_apply :
+  hasse_deriv k f = f.sum (λ i r, monomial (i - k) (↑(i.choose k) * r)) := rfl
+
+lemma hasse_deriv_coeff (n : ℕ) :
+  (hasse_deriv k f).coeff n = (n + k).choose k * f.coeff (n + k) :=
+begin
+  rw [hasse_deriv_apply, coeff_sum, sum_def, finset.sum_eq_single (n + k), coeff_monomial],
+  { simp only [if_true, nat.add_sub_cancel, eq_self_iff_true], },
+  { intros i hi hink,
+    rw [coeff_monomial],
+    by_cases hik : i < k,
+    { simp only [nat.choose_eq_zero_of_lt hik, if_t_t, nat.cast_zero, zero_mul], },
+    { push_neg at hik, rw if_neg, contrapose! hink, exact (nat.sub_eq_iff_eq_add hik).mp hink, } },
+  { intro h, simp only [not_mem_support_iff.mp h, monomial_zero_right, mul_zero, coeff_zero] }
+end
+
+lemma hasse_deriv_zero' : hasse_deriv 0 f = f :=
+by simp only [hasse_deriv_apply, nat.sub_zero, nat.choose_zero_right,
+  nat.cast_one, one_mul, sum_monomial_eq]
+
+@[simp] lemma hasse_deriv_zero : @hasse_deriv R _ 0 = linear_map.id :=
+linear_map.ext $ hasse_deriv_zero'
+
+lemma hasse_deriv_one' : hasse_deriv 1 f = derivative f :=
+by simp only [hasse_deriv_apply, derivative_apply, monomial_eq_C_mul_X, nat.choose_one_right,
+    (nat.cast_commute _ _).eq]
+
+@[simp] lemma hasse_deriv_one : @hasse_deriv R _ 1 = derivative :=
+linear_map.ext $ hasse_deriv_one'
+
+@[simp] lemma hasse_deriv_monomial (n : ℕ) (r : R) :
+  hasse_deriv k (monomial n r) = monomial (n - k) (↑(n.choose k) * r) :=
+begin
+  ext i,
+  simp only [hasse_deriv_coeff, coeff_monomial],
+  by_cases hnik : n = i + k,
+  { rw [if_pos hnik, if_pos, ← hnik], apply nat.sub_eq_of_eq_add, rwa add_comm },
+  { rw [if_neg hnik, mul_zero],
+    by_cases hkn : k ≤ n,
+    { rw [← nat.sub_eq_iff_eq_add hkn] at hnik, rw [if_neg hnik] },
+    { push_neg at hkn, rw [nat.choose_eq_zero_of_lt hkn, nat.cast_zero, zero_mul, if_t_t] } }
+end
+
+lemma hasse_deriv_C (r : R) (hk : 0 < k) : hasse_deriv k (C r) = 0 :=
+by rw [← monomial_zero_left, hasse_deriv_monomial, nat.choose_eq_zero_of_lt hk,
+    nat.cast_zero, zero_mul, monomial_zero_right]
+
+lemma hasse_deriv_apply_one (hk : 0 < k) : hasse_deriv k (1 : polynomial R) = 0 :=
+by rw [← C_1, hasse_deriv_C k _ hk]
+
+lemma hasse_deriv_X (hk : 1 < k) : hasse_deriv k (X : polynomial R) = 0 :=
+by rw [← monomial_one_one_eq_X, hasse_deriv_monomial, nat.choose_eq_zero_of_lt hk,
+    nat.cast_zero, zero_mul, monomial_zero_right]
+
+lemma factorial_smul_hasse_deriv :
+  ⇑(k! • @hasse_deriv R _ k) = ((@derivative R _)^[k]) :=
```

Comment:
Instead of `^[_]`, doesn't plain `^` work?

Reply:
`^[_]` is what `src/data/polynomial/derivative.lean` uses all over the place.